src/scout_apm/core/thread_local.py

```diff
@@ -4,23 +4,37 @@
 
 
 class ThreadLocalSingleton(object):
+    _instance_lock = threading.Lock()
+
     @classmethod
     def instance(cls, *args, **kwargs):
-        if not hasattr(cls, '_thread_lookup'):
-            cls.__new_instance(args, kwargs)
-        elif not hasattr(cls._thread_lookup, 'instance'):
-            cls.__new_instance(args, kwargs)
-        elif cls._thread_lookup.instance is None:
-            cls.__new_instance(args, kwargs)
-        return cls._thread_lookup.instance
+        with cls._instance_lock:
+            # The thread local dict doesn't exist at all. Need to create both it and the instance
+            if not hasattr(cls, '_thread_lookup'):
+                cls.__new_instance(args, kwargs)
+                return cls._thread_lookup.instance
+
+            # Somehow we have the thread local dict, but failed to make the instance. Just reinit both
+            if not hasattr(cls._thread_lookup, 'instance'):
+                cls.__new_instance(args, kwargs)
+                return cls._thread_lookup.instance
+
+            # The instance is none, so we'll need to recreate it (and the thread local dict too)
+            if cls._thread_lookup.instance is None:
+                cls.__new_instance(args, kwargs)
+                cls._thread_lookup.instance
+
+            return cls._thread_lookup.instance
```

Comment:
Either be consistent and return in every `if` case, and log an error if you reach here. Or don't return in the `if` scopes and instead only one return at the end.